extendr-macros/src/lib.rs

```diff
@@ -168,3 +169,43 @@ pub fn R(item: TokenStream) -> TokenStream {
 pub fn Rraw(item: TokenStream) -> TokenStream {
     R::R(item.into(), false).into()
 }
+
+/// Derives an implementation of `From<Struct> for Robj` and `TryFrom<Robj> for Struct` on this struct.
+/// A named list is used as the R version of the struct, where the where the list names correspond to the field names of the Rust struct.
+///
+/// # Examples
+/// ```ignore
+/// use extendr_api::prelude::*;
+/// use extendr_macros::ConvertRList;
+///
+/// #[derive(ConvertRList)]
+/// struct Foo {
+///     a: usize,
+///     b: String
+/// }
+/// let robj: Robj = Foo {
+///     a: 5,
+///     b: String::from("bar")
+/// }.into();
+/// ```
+/// Will roughly convert to:
+/// ```r
+/// list(
+///     a = 5,
+///     b = "bar"
+/// )
+/// ```
+/// You can also create an instance of your struct from R:
+/// ```ignore
+/// let foo: Foo = R!("list(a = 5, b = 'bar')").unwrap().try_into().unwrap();
+/// ```
+/// # Details
+/// Note, the `From<Struct> for Robj` behaviour is different from what you get by applying the standard `#[extendr]` macro
+/// to an `impl` block. The `#[extendr]` behaviour returns to R a **pointer** to Rust memory, and generates wrapper functions for calling
+/// Rust functions on that pointer. The implementation from `#[derive(ConvertRList)]` actually converts the Rust structure
+/// into a native R list, which allows you to manipulate and access the internal fields, but it's a one-way conversion,
+/// and converting it back to Rust will produce a copy of the original struct.
+#[proc_macro_derive(ConvertRList)]
```

Comment:
We should have to separate derives.

I would suggest `ToList, ToRobj, FromList and FromRobj`

The List versions use `List` instead of `Robj`.

We may want to implement some of the traits manually.

Reply:
What is the advantage of a whole trait to convert to/from `List` when you can get one in one function call from a `Robj`?

Reply:
In the latest commit I have separated the derives, although I have not made a separate trait for `List`s.